### ISSUE TITLE
[RFC] Package subpath exports for exporting internals

### ISSUE THREAD

**Issue by Rugvip** [backstage/backstage#7926]
**Status:** Open for comments

## Need

The problem that triggered this RFC is the uncertainties around what a Backstage plugin package should export and how to maintain these exports over time.

Both backend and frontend plugins typically export a set of high-level building blocks that make up the public API. For a frontend package these would be things like the plugin instance, extensions, Utility APIs and their implementations. It's a bit less structured for backend packages, but there will typically be a `createRouter` function along with a builder or a couple of high-level interfaces and implementations that help construct the router. All of these exports are generally things that we want to keep stable.

In addition to these high-level APIs it has become increasingly common that a plugin package exports a second set of APIs, ones that typically where internal to begin with, but ended up needing to be exported for increased customizability. For frontend plugins this generally tends to be React components, and sometimes hooks, while in the backend world it is not as well defined and tends to be more of a gradual scale of increasingly low-level or more internal concerns.

These exports should not be confused by the ones from `-react`, `-node`, or `-common` plugin packages such as `@backstage/plugin-catalog-react`, as these packages are intended to provide something that is more akin to a library. What we're considering here are exports from a plugin package that don't need to be used for a standard setup of the plugin, but only for more advanced usages such as internal wrappers or custom architectures.

To get a bit more concrete this is a list of pull requests that add these kind of exports. I feel I need to point out that there is obviously nothing wrong with these pull requests, as this is the pattern we have in place at the moment.

- [Export TemplateList from Scaffolder plugin](https://github.com/backstage/backstage/pull/7755)
- [Export DomainCard from Explore plugin](https://github.com/backstage/backstage/pull/7277)
- [Multiple Scaffolder backend exports](https://github.com/backstage/backstage/pull/7426)
- [Export internals from Git Release Manager](https://github.com/backstage/backstage/pull/6922)

... and there are many more ...

I think there are a couple of ways that this hurts plugins. The most important one I believe is discoverability. It makes it difficult to look at a set of package exports and understand what they do and which ones you should use. I think this is especially important for frontend plugins where there *could* be a cleaner surface of a few exported extensions, a plugin instance, and the rest is related to Utility APIs. It makes auto-completion in editors less useful, and API reference documentation becomes more cluttered.

The other very important aspect is maintainability of plugins, which of course is the direction I am coming from :grin:. A larger API surface is more difficult to maintain and evolve over time. It causes more work to be waiting on deprecation periods and more code hanging around for backwards compatibility. This is something that of course affects the maintainer team, but also anyone that wants to maintain a high quality and stable plugin, both ones that are open source and internal to an organization.

## Proposal

Something that I've realized might be a solution to this problem are additional entry points on a different path in the `exports` field in `package.json`, also known as [subpath exports](https://github.com/nodejs/node/blob/master/doc/api/packages.md#subpath-exports). There are plenty of libraries that already use this or something like it for various purposes. One example is MUI, where you can import things both from the root `@material-ui/core`, but also `@material-ui/core/styles`, or `@material-ui/core/Button`. Now the implementation isn't quite the same as they don't use the `exports` field, but we'll call that an implementation detail :grin:

The way that I propose that we use these exports to solve the described problem is to treat the main entrypoint as the public API, for example `@backstage/plugin-catalog`. In addition to this, a package is able to declare additional entry points using subpath exports, for example `@backstage/plugin-catalog/components`. These entry points would always be considered internal API, and unless stated otherwise would not be subject to the API stability and [Semver](https://semver.org/) rules as the main entrypoint. The exact paths used would be something to be discussed, but it could be a mix of `./components`, `./hooks` and so on for frontend plugins, but we might also settle on just using `./internal` for everything.

If this is something that we end up wanting to move forward there will be plenty of places in the build tooling where we need to add support for this pattern, and of course existing packages to refactor. The good news is that Node.js now has a standardized way to specify subpath exports, so we wouldn't be building a custom solution but rather build out support for something that becoming part of the ecosystem anyway.

## Alternatives

There are plenty of different ways that we could utilize subpath exports, particularly with different patterns as mentioned above. There are also different ways that we could handle additional entry points that should be considered main entry points under Semver. There's also a question whether we should be generating API reports for all of the entry points (I think we should, they're really useful) and also if and how we generate API reference documentation.

Apart from all of these alternatives for how to apply subpath exports, there are of course also other alternatives. Perhaps the primary one that has come up is a more documentation driven approach where we mark certain APIs as `@internal` and try to highlight that in documentation. I think that this is very error prone though, both in the producing and the consuming end. It's easy to forget to mark something as internal, and it's easy to use something internal by mistake. It also doesn't solve the discoverability issue as everything is exposed through the same pile of everything in the main entrypoint. It is much more obvious that `import { FooService } from 'backstage-plugin-foo/internal'` is importing something internal compared to `import { FooService } from 'backstage-plugin-foo'`.

## Risks

I think one of the biggest risks if we put this all in place is that the `./internal` entry point (or whatever we'd choose) becomes a dumping ground and a bad experience for anyone consuming it. That's why I'd still suggest that we generate API reports and documentation for all entry points and not make it a complete wild west. I also think changes to any of the entry points should be listed in the changelog.

Perhaps the biggest risk is the built tooling complexity though. While the `exports` field is standardized not all tooling is there yet, and we might need to come up with a couple of magic solutions to make it all work.

**Comment by benjdlambert** [comment]
Upon further investigation it appears that there could be an alternative approach in order to do this.

For local development, we can use both the `package.json#exports` field and `package.json#typesVersions` field in order to create the typings and references to the different entrypoints, and then have our `backstage-cli build` command create the the entrypoints and their typings in the `dist` folder and re-write the `package.json`s with the correct type definitions.

Tested out locally and this seems to work:

```json
  "typesVersions": {
    "*": {
      "*": [
        "src/index.ts"
      ],
      "alpha": [
        "src/alpha.ts"
      ]
    }
  },
  "exports": {
    ".": "./src/index.ts",
    "./alpha": "./src/alpha.ts"
  },
  ```

Hopefully when we have the new `moduleResolution: bundler` syntax available in `typescript@v5` we may no longer need the `typesVersions` reference we can embed everything in the `exports` field without using `type: module` in `package.json` 🤞

**Comment by webark** [comment]
is this mainly being tracked here, or here https://github.com/backstage/backstage/issues/12218

**Comment by Rugvip** [comment]
@webark this is actually completely closed as implemented through #15816 😁. There's still an evolution awaiting for packages to potentially start using this more, but I feel it's best to have one or two release cycles to make sure it's all solid before we encourage using it for all kinds of things.
